perp/man/tinylog.8
.\" tinylog.8
.\" wcm, 2009.09.29 - 2011.02.01
.\" ===
.TH tinylog 8 "March 2011" "perp-2.04" "persistent process supervision"
.SH NAME
tinylog \- log stdin to a directory of rotated log files
.SH SYNOPSIS
.B tinylog [\-hV] [\-k
.I numkeep
.B ] [\-r] [\-s
.I logsize
.B ] [\-t] [\-z]
.I dir
.SH DESCRIPTION
.B tinylog
reads lines from standard input and writes them to a set of rotated log files
maintained in
.IR dir . 
.PP
While active,
.B tinylog
switches into
.I dir
and writes into a log file named
.IR current .
When
.I current
reaches the size specified with the
.B -s
option (default 100000 bytes),
it rotates the file:
.I current
is renamed with a filename in the form
.IR _yyyymmddThhmmss.uuuuuu.s ,
and a new
.I current
is opened.
.PP
.B tinylog
maintains a set of older rotated log files in
.IR dir ,
upto the number specified with the
.B -k
option (default 5).
After this number is reached,
.B tinylog
deletes the rotated log file with the oldest timestamp before
continuing with a new
.IR current .
.PP
If the
.B -z
option is specified,
.B tinylog
will compress log files during rotation with the utility specified in the
environmental variable
.B TINYLOG_ZIP
(default
.IR /usr/bin/gzip ).
If the compression is successful,
the rotated log file is renamed with a
.I .Z
extension.
.PP
The name of a rotated log file may be described further:
beginning with an underscore,
followed by a current
.B gmtime(3)
timestamp in RFC8601 format (to the nearest microsecond),
followed by a status suffix, 
and optionally followed by a
.I .Z
zip extension.
Normally the status suffix is
.IR .s ,
indicating the file was safely written to disk.
Otherwise a status suffix of
.I .u
indicates that the file was not safely written to disk,
and may possibly be incomplete and/or corrupted by an unexpected failure.
.PP
.B tinylog
sets the file mode of
.I current
to 0644 while active.
When
.B tinylog
sees eof on stdin,
it writes any pending line to
.IR current ,
fsyncs and closes the file,
and changes its mode to 0744 before exiting.
Whenever
.B tinylog
restarts,
it looks for an existing
.I current
and,
if one is found with a file mode 0744,
reopens it and changes its mode back to 0644 for continued logging.
Otherwise,
if an existing
.I current
is found set with a mode of 0644,
it is immediately rotated with a
.I .u
suffix,
and a new
.I current
is opened.
.PP
.B tinylog
ignores empty lines,
truncates lines longer than 1000 characters,
and converts unprintable control characters to `?'.
.SH OPTIONS
.TP
.B \-h
Help.
Print a brief usage message to stderr and exit.
.TP
.B \-k numkeep
Keep.
Sets the maximum number of log files that
.B tinylog
will keep after rotation.
Whenever
.B tinylog
rotates the most recent log file,
it deletes any more than
.I numkeep
older log files found in the directory.
The minimum number is 0,
in which case no older log files are kept.
If not specified,
the default number of older log files kept is 5.
.TP
.B \-r
Rotate on start.
Normally on start-up,
.B tinylog
begins logging with an existing
.I current
file if it has been safely closed from a previous session.
The
.B \-r
option causes
.B tinylog
to immediately rotate any existing
.I current
file and begin logging with a new one.
.TP
.B \-s logsize
Size.
Sets the maximum size (in bytes) that a log file may grow before
rotation.
The minimum size is 2000.
The default size is 100000.
.TP
.B -t
Timestamp.
A current
.B gmtime(3)
timestamp string in the form of ``yyyymmddThhmmss.uuuuuu'' is prepended
to each line written to the log file.
.TP
.B \-V
Version.
Print the version number to stderr and exit.
.TP
.B \-z
Zip.
This option instructs
.B tinylog
to run a compression utility when it rotates the log file.
The compression utility may be specified in the environmental variable
.BR TINYLOG_ZIP .
If the variable is not set or empty,
.B tinylog
will use a compiled-in default,
usually
.IR /usr/bin/gzip .
The compression utility should be designed to run without any arguments,
reading from stdin and writing to stdout.
After successful compression,
.B tinylog
will rename the rotated log file with a
.I .Z
extension.
.SH ENVIRONMENT
TINYLOG_ZIP
.RS
If defined and the
.B \-z
option is specified,
will be taken as the executable for the compression utility to use when rotating a log file.
.RE
.SH SIGNALS
.B tinylog
traps the following signals for adminstrative control during runtime:
.PP
SIGHUP
.RS
Close and rotate
.IR current ,
then continue logging with a new
.IR current .
.RE
.PP
SIGTERM
.RS
Stop reading stdin,
process any pending lines already read,
flush and close
.IR current
safely to disk,
then exit 0 (no error).
Stdin will be left at the first byte of any unprocessed data.
.RE
.SH EXIT STATUS
.B tinylog
exits with the following values:
.TP
0
Normally
.B tinylog
runs until eof is found on stdin,
or until it receives a
.B TERM
signal.
In these cases,
.B tinylog
writes any pending lines to the log,
flushes and closes
.IR current ,
changes its mode to 0744,
and exits 0.
.TP
100
.B tinylog
failed on startup because of some usage error,
such as invalid command-line option or missing argument.
In this case
.B tinylog
will print a brief diagnostic to stderr on exit.
.TP
111
.B tinylog
found another instance or active lockfile running in
.IR dir ,
or failed due to some other system or resource error on startup.
In this case
.B tinylog
will print a brief diagnostic to stderr on exit.
.PP
Otherwise,
after startup,
.B tinylog
tries very hard not to exit until eof is read on stdin.
In the case of system or resource failures,
.B tinylog
will print a diagnostic message to stderr,
pause for a brief interval,
and then retry the failed operation.
.SH AUTHOR
Wayne Marshall, http://b0llix.net/perp/
.SH SEE ALSO
.nh
.BR perp_intro (8),
.BR perpboot (8),
.BR perpctl (8),
.BR perpd (8),
.BR perpetrate (5),
.BR perphup (8),
.BR perpls (8),
.BR perpok (8),
.BR perpstat (8),
.BR sissylog (8)
.\" EOF tinylog.8


CHANGES
# CHANGES
# perp release history and changelog
# wcm, 2010.02.17 - 2011.03.22
# ===

perp-2.04 (2011.03.22):
 * perphup:
   - added option -t, trigger perpd shutdown
 * perpboot:
   - waitpid() for deux[1] on SIGTERM
 * perp-setup:
   - added arg for ctldir
   - configure symlink /etc/perp/.control --> ctldir
   - configure rc.perp bootscript for runchoom(8)
 * runchoom:
   - new, utility for linux "oom killer" abatement

perp-2.03 (2011.03.11):
 * perp-setup:
   - in rc.perp boot script for perpd(8), setup default runlimit(8)
     configuration to hard limits for RLIMIT_NPROC, RLIMIT_NOFILE

perp-2.02 (2011.02.22):
 * perpd:
   - eliminated some spurious/redundant client connection logging

perp-2.01 (2011.02.11):
 * perp-setup:
   - runlimit added to .boot/rc.perp script
 * perpd:
   - added option -g, enable group permission on control socket
   - additional comments in LIMITS section of perpd(8) manual
 * runlimit:
   - added option -E, take resource limits from environment
   - added option -F, take resource limits from file
 * lasagna:
   - outvec/mk_outvec.c: portability fix for dietlibc

perp-2.00 (2011.02.02):
 * perpd:
   - coalesce supervisor (formerly perpetrate(8)) into scanner/controller
   - runs as a single process
   - all ipc with client applications via single domain socket 
   - ipc protocol version 2 (see PROTO_V2.txt)
 * perpetrate:
   - eliminate perpetrate(8) executable (subsumed by perpd(8))
   - perpetrate(5) manual page retained for runscript conventions
 * perpctl:
   - add commands [A | X], activate/deactivate service
 * libasagna:
   - nuscan.h, improve numeric scanning interface (supercedes nscan.h)
   - other iterative improvements incorporated from projects hdb and ezcdb
 
perp-0.00 (2010.01.12):
 - maiden public release

### EOF


perp/tinylog.c
/* tinylog.c
** log from data read on stdin
** wcm, 2006.08.04 - 2009.12.09
** ===
*/

/* TODO (maybe ...):
**   USR1/USR2 toggle fsync() on each ouput line
**   provide short-circuit logic for keep if keep_max == 0
**   provide unlimited keep files
**   provide unlimited growth of current if size == 0
*/

/* standard libs: */
#include <stdlib.h>
/* rename() from stdio.h: */
extern int     rename(const char *oldpath, const char *newpath);

/* unix libs: */
#include <unistd.h>
#include <dirent.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <time.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/wait.h>

/* lasagna: */
#include "uchar.h"
#include "buf.h"
#include "cstr.h"
#include "fd.h"
#include "ioq.h"
#include "nextopt.h"
#include "nfmt.h"
#include "nscan.h"
#include "pidlock.h"
#include "sig.h"
#include "sigset.h"
#include "sysstr.h"
#include "tain.h"

/* environ: */
extern char  **environ;

/* local: */
#define  EPUTS_IOQ 1
#include "eputs.h"
#include "loggit.h"
#include "tinylog.h"

/* logging variables in scope: */
static const char *progpid = NULL;
static const char *progname = NULL;
static const char prog_usage[] =
    " [-hV] [-k numkeep] [-r] [-s logsize] [-t] [-z] dir";

/*
** configure loggit.h macros:
*/
#define LOGGIT_FMT \
  progname, "[", progpid, "]: "


/*
** tinylog object:
*/
struct tinylog {
   const char    *fn_logdir;
   int            fd_orig;
   int            fd_logdir;
   int            fd_pidlock;
   int            fd_current;
   size_t         current_size;
   size_t         current_max;
   size_t         keep_max;
   int            wantstamp;
   int            wantzip;
};


/* ioq for stdin: */
#define INBUF_SIZE  1024
uchar_t        inbuf[INBUF_SIZE];
ssize_t        read_op(int fd, void *buf, size_t len);
ioq_t          in = ioq_INIT(0, inbuf, sizeof inbuf, &read_op);

/*
** variables in scope:
*/
static pid_t   mypid = 0;
static int     flagexit = 0;
static int     flagrotate = 0;

/* sigset for blocking/unblocking signal handler: */
static sigset_t my_sigset;

/* global timestamp string: */
char           stamp8601[] = "yyyymmddThhmmss.uuuuuu";
/* global filename string for log archive: */
char           fn8601[] = "_yyyymmddThhmmss.uuuuuu.?" ZIP_EXT;

/* where to find gzip: */
const char    *gzip_path = NULL;

/* default maximum size for log file (bytes): */
#define CURRENT_MAX  100000
/* maximum size for line in log file (bytes): */
#define LOGLINE_MAX    1000

/* RETRY():
**  if test evaluates true: issue warning, pause, and repeat
**  test is a system call that sets errno
**  exits failure on sigterm
**  example usage:
**      RETRY(chdir(".") == -1),
**          "fail chdir() on current directory");
*/
#define RETRY(test, ...) \
  {\
      while((test)){ \
          tain_t  t = tain_INIT(2,0); \
          if(flagexit){ \
              fatal_syserr("exiting on SIGTERM during retry error: ", __VA_ARGS__); \
          } \
          warn_syserr("pausing: ", __VA_ARGS__); \
          tain_pause(&t, NULL); \
      } \
  }


/*
** declarations in scope:
*/
static int     write_all(int fd, void *buf, size_t len);
static void    stamp8601_make(char *stamp8601);
static void    init_logdir(struct tinylog *tinylog);
static void    init_current(struct tinylog *tinylog, int resume);
static void    tinylog_rotate(struct tinylog *tinylog);
static void    tinylog_keep(struct tinylog *tinylog, const char *filename,
                            const char *ext);
static int     tinylog_prune(struct tinylog *tinylog);
static int     tinylog_gzip(struct tinylog *tinylog, const char *file);
static void    tinylog_post(struct tinylog *tinylog, char *logline,
                            size_t len);
static int     do_log(struct tinylog *tinylog);


/*
** definitions:
*/

static
void sig_handler(int sig)
{
   switch (sig) {
   case SIGTERM:
      ++flagexit;
      break;
   case SIGHUP:
      ++flagrotate;
      break;
   default:
      break;
   }

   return;
}


/* read_op()
**   read() operation installed for ioq on stdin
*/
ssize_t read_op(int fd, void *buf, size_t len)
{
   ssize_t        r = 0;

   /* note:
    **
    **   SIGTERM handler setup *without* SA_RESTART sigaction flag
    **   so SIGTERM will interrupt the read() here
    */
   do {
      /* if SIGTERM, simulate eof before read(): */
      if (flagexit)
         return 0;
      r = read(fd, buf, len);
   } while ((r == -1) && (errno == EINTR));

   return r;
}


static
int write_all(int fd, void *buf, size_t len)
{
   ssize_t        w = 0;

   while (len) {
      do {
         w = write(fd, buf, len);
         if ((w == -1) && (errno == EAGAIN)) {
            int            terrno = errno;
            tain_t         t = tain_INIT(1, 0);
            warn_syserr("pausing for write() error");
            tain_pause(&t, NULL);
            errno = terrno;
         }
      } while ((w == -1) && ((errno == EINTR) || (errno == EAGAIN)));

      if (w == -1)
         return -1;             /* error! */
      if (w == 0);              /* XXX, no bytes written (shouldn't happen?) */

      buf += w;
      len -= w;
   }

   /* return:
    **    = -1, error (above)
    **    =  0, success (here)
    */

   return 0;
}


/* stamp8601_make()
** generate an rfc8601-type ascii timestamp into buffer stamp
**
** notes:
**   caller must supply stamp of sufficient size: (sizeof stamp8601)
**   stamp is not nul-terminated
*/
static
void stamp8601_make(char *stamp)
{
   char          *s = stamp;
   tain_t         now;
   time_t         now_utc;
   struct tm     *tm_now = NULL;

   tain_now(&now);
   now_utc = tain_to_utc(&now);
   tm_now = gmtime(&now_utc);

   nfmt_uint32_pad0_(s, 1900 + tm_now->tm_year, 4);
   s += 4;
   /* use mday format (day of month): */
   nfmt_uint32_pad0_(s, 1 + tm_now->tm_mon, 2);
   s += 2;
   nfmt_uint32_pad0_(s, tm_now->tm_mday, 2);
   s += 2;
   *s = 'T';
   ++s;
   nfmt_uint32_pad0_(s, tm_now->tm_hour, 2);
   s += 2;
   nfmt_uint32_pad0_(s, tm_now->tm_min, 2);
   s += 2;
   nfmt_uint32_pad0_(s, tm_now->tm_sec, 2);
   s += 2;
   *s = '.';
   ++s;
   nfmt_uint32_pad0_(s, (uint32_t) (now.nsec / 1000), 6);

   return;
}


static
void init_logdir(struct tinylog *tinylog)
{
   int            fd;

   /* open log directory: */
   mkdir(tinylog->fn_logdir, 0700);
   if (chdir(tinylog->fn_logdir) == -1) {
      fatal_syserr("failure chdir() to log directory ", tinylog->fn_logdir);
   }

   if ((fd = open(".", O_RDONLY)) == -1) {
      fatal_syserr("failure open() on log directory ", tinylog->fn_logdir);
   }
   fd_cloexec(fd);
   tinylog->fd_logdir = fd;

   if ((fd = pidlock_set(PIDLOCK, mypid, PIDLOCK_NOW)) == -1) {
      fatal_syserr("failure acquiring lock file ", PIDLOCK,
                   " in log directory ", tinylog->fn_logdir);
   }
   fd_cloexec(fd);
   tinylog->fd_pidlock = fd;

   return;
}


/* init_current()
**   setup or resume "current" logfile
**   if resume == 0: open new current
**   on entry and exit, cwd is logging directory fd_logdir
*/
static
void init_current(struct tinylog *tinylog, int resume)
{
   struct stat    sb;
   int            e;
   int            fd;
   int            new = 0;

   /* stat current: */
   e = stat("current", &sb);
   if (e == -1) {
      if (errno != ENOENT) {
         /* current exists but stat() failed: */
         fatal_syserr("failure stat() on current in log directory");
      } else {
         /* current does not exist, will be new: */
         ++new;
      }
   } else {
      if (!(sb.st_mode & 0100)) {
         /* current exists, found in mode 0644: last exit not clean: */
         tinylog_keep(tinylog, "current", "u");
         /* current will now be new: */
         ++new;
      } else if (resume == 0) {
         /* current exists with clean exit status, but new current requested: */
         tinylog_keep(tinylog, "current", "s");
         /* current will now be new: */
         ++new;
      }
      /* otherwise: current exists with clean exit status and will be resumed */
   }

   /* whether current exists or will be new, open() it: */
   fd = open("current", O_WRONLY | O_CREAT | O_APPEND | O_NONBLOCK, 0600);
   if (fd == -1) {
      fatal_syserr("failure open() on current in log directory");
   }

   /* mode for active logging: */
   if (fchmod(fd, 0644) == -1) {
      fatal_syserr("failure fchmod() on current in log directory");
   }
   fd_cloexec(fd);

   /* all set: */
   tinylog->fd_current = fd;
   tinylog->current_size = new ? 0 : sb.st_size;

   return;
}


/* tinylog_rotate()
**   flush current to disk;
**   rename as previous; archive previous;
**   open new current
*/
static
void tinylog_rotate(struct tinylog *tinylog)
{
   int            fd = tinylog->fd_current;

   if (!(tinylog->current_size > 0)) {
      flagrotate = 0;
      return;
   }

   fsync(fd);
   close(fd);

   RETRY((rename("current", "previous") == -1),
         "failure rename() on current");

   tinylog_keep(tinylog, "previous", "s");

   RETRY(((fd =
           open("current", O_WRONLY | O_CREAT | O_APPEND | O_NONBLOCK,
                0600)) == -1), "failure open() for new current");

   /* mode for active logging: */
   RETRY((fchmod(fd, 0644) == -1), "failure fchmod() for new current");
   fd_cloexec(fd);

   /* all set: */
   tinylog->fd_current = fd;
   tinylog->current_size = 0;
   flagrotate = 0;

   return;
}


/* tinylog_keep()
**   rotate log to timestamped archive with extension
**   on entry and exit, cwd is logdir
*/
static
void tinylog_keep(struct tinylog *tinylog, const char *log, const char *ext)
{
   struct stat    sb;
   tain_t         t;
   int            e;

   for (;;) {
      e = stat(log, &sb);
      if (e == 0)
         break;
      if ((e == -1) && (errno == ENOENT))
         return;
      /* else: stat() error: */
      warn_syserr("pausing: failure stat() for log ", log);
      tain_LOAD(&t, 2, 0);
      tain_pause(&t, NULL);
      continue;
   }

   if (sb.st_nlink == 1) {
      fn8601[0] = '_';
      fn8601[23] = '.';
      fn8601[24] = ext[0];
      fn8601[25] = '\0';

      for (;;) {
         stamp8601_make(&fn8601[1]);
         e = link(log, fn8601);
         if (e == 0)
            break;
         if ((e == -1) && (errno == EEXIST)) {
            /* collision on microsecond timestamp: */
            tain_LOAD(&t, 0, 2000);
            tain_pause(&t, NULL);
            continue;
         }
         /* else link() error: */
         warn_syserr("pausing: failure link() for ", log, " to ", fn8601);
         tain_LOAD(&t, 2, 0);
         tain_pause(&t, NULL);
         continue;
      }
   }

   RETRY((unlink(log) == -1), "failure unlink() on file ", log);

   if (tinylog->wantzip) {
      e = tinylog_gzip(tinylog, fn8601);
      if (e != 0) {
         log_warning("failure gzip on log archive ", fn8601);
      }
   }

   while (tinylog_prune(tinylog) == -1) {
      log_warning("pausing: failure while pruning log directory");
      tain_LOAD(&t, 2, 0);
      tain_pause(&t, NULL);
      continue;
   }

   /* success: */
   return;
}


/* tinylog_prune()
**   scan logdir and prune oldest log archive
**   return
**     1 : an archive was pruned, yet more remain to be pruned
**     0 : logdir pruned (either nothing to prune or prune complete)
**    -1 : system error from opendir(), readdir(), or unlink(), errno set
**
**   on entry and exit, cwd is logdir
*/
static
int tinylog_prune(struct tinylog *tinylog)
{
   DIR           *dir;
   struct dirent *d;
   char           target[80] = "_z";
   size_t         n;
   int            terrno;
   int            keep = tinylog->keep_max;
   int            count = 0;

   if ((dir = opendir(".")) == NULL) {
      warn_syserr("failure opendir(\".\") in log directory");
      return -1;
   }

   /* repeat prune operation while count > keep: */
   do {
      /* scan: find "oldest" archive file (from its filename) */
      for (;;) {
         errno = 0;
         if ((d = readdir(dir)) == NULL) {
            /* done or failure: */
            break;
         }
         if ((d->d_name[0] == '_')
             && (n = cstr_len(d->d_name) >= 23)
             && (d->d_name[9] == 'T')
             && (n < sizeof target)) {
            /* smells like a log archive: */
            ++count;
            /* target: oldest filename found in scan: */
            if (cstr_cmp(d->d_name, target) < 0) {
               cstr_copy(target, d->d_name);
            }
         }
      }
      terrno = errno;
      closedir(dir);

      if (terrno) {
         /* oops, readdir() error in loop: */
         errno = terrno;
         warn_syserr("failure readdir() while scanning log directory");
         return -1;
      }

      if (count <= keep) {
         /* nothing to prune: */
         return 0;
      }

      /* remove the oldest log archive in scan: */
      if (unlink(target) == -1) {
         warn_syserr("failure unlink() to prune target log ", target);
         return -1;
      }

      /* target archive pruned: */
      --count;

   } while (count > keep);

   return 0;
}


/* tinylog_gzip()
**   execve() a gzip child process: file -> "zipped"
**   on successful completion of child:
**     rename() "zipped" -> file + ZIP_EXT
**     unlink() original file
**
**   return
**     0:  success
**    -1:  something failed
*/
static
int tinylog_gzip(struct tinylog *tinylog, const char *file)
{
   char           gzipfile[80] = "";
   pid_t          pid;
   int            wstat;

   RETRY(((pid = fork()) == -1), "fail fork() for gzip process");

   if (pid == 0) {              /* child */
      int            fd;
      const char    *args[2];

      sig_uncatch(SIGTERM);
      sig_uncatch(SIGHUP);
      sigset_unblock(&my_sigset);

      if ((fd = open(file, O_RDONLY | O_NONBLOCK)) == -1) {
         return -1;
      }
      fd_move(0, fd);

      if ((fd =
           open("zipped", O_WRONLY | O_CREAT | O_TRUNC | O_NONBLOCK,
                0600)) == -1) {
         return -1;
      }
      fd_move(1, fd);

      args[0] = gzip_path;
      args[1] = NULL;
      execve(gzip_path, (char *const *) args, environ);
      fatal_syserr("fail execve() for gzip process");
   }

   /* parent */
   RETRY((waitpid(pid, &wstat, 0) == -1), "fail waitpid() for gzip process");

   if (WIFEXITED(wstat) && (WEXITSTATUS(wstat) != 0)) {
      char           nbuf[NFMT_SIZE];
      nfmt_uint32(nbuf, WEXITSTATUS(wstat));
      log_warning("gzip process exited non-zero error code: ", nbuf);
      unlink("zipped");
      return -1;
   }
   if (WIFSIGNALED(wstat)) {
      log_warning("gzip process terminated on signal: ",
                  sysstr_signal(WTERMSIG(wstat)));
      unlink("zipped");
      return -1;
   }

   cstr_vcat(gzipfile, file, ZIP_EXT);
   if (rename("zipped", gzipfile) == -1) {
      warn_syserr("fail rename() after gzip");
      unlink("zipped");
      return -1;
   }

   if (chmod(gzipfile, 0644) == -1) {
      warn_syserr("fail chmod() after gzip");
      unlink(gzipfile);
      return -1;
   }

   /* archive zipped successfully, delete original file: */
   RETRY((unlink(file) == -1), "fail unlink() after gzip on ", file);

   return 0;
}


static
void tinylog_post(struct tinylog *tinylog, char *logline, size_t len)
{
   int            e;

   /* rotate? */
   if (flagrotate || (tinylog->current_size >= (tinylog->current_max - len))) {
      tinylog_rotate(tinylog);
   }

   /* post logline: */
   e = write_all(tinylog->fd_current, logline, len);
   if (e != 0) {
      /* XXX */
      warn_syserr("write() error on current");
      return;
   }
#if 0
   /* XXX, make this an option? */
   fsync(tinylog->fd_current);
#endif

   tinylog->current_size += len;

   return;
}


static
int do_log(struct tinylog *tinylog)
{
   char           logline[LOGLINE_MAX + 1];
   size_t         len = 0;
   char           c;
   int            r;
   int            wantstamp = tinylog->wantstamp;
   size_t         startpos = (wantstamp ? 24 : 0);
   int            eof = 0;

   /* terminal condition: eof */
   /* outer loop: one line per loop: */
   while (!eof) {
      /* initialize line: */
      len = startpos;
      /* inner loop: one char per loop: */
      while ((r = ioq_GETC(&in, &c)) == 1) {
         if (c == '\n') {
            /* newline appended after break (if line non-empty): */
            break;
         }
         if (len < LOGLINE_MAX) {
            /* filter unprintable control chars: */
            if ((c < 32) && (c != '\t'))
               c = '?';
            logline[len++] = c;
         }
         /* implicit else: draining input buffer from logline overflow */
      }
      if (r == -1)              /* XXX, io error on stdin: */
         return -1;
      if (r == 0)
         ++eof;
      /* post any non-empty line collected from inner loop: */
      if (len > startpos) {
         /* prepend timestamp: */
         if (wantstamp) {
            stamp8601_make(logline);
            logline[22] = ':';
            logline[23] = ' ';
         }
         /* append newline: */
         logline[len++] = '\n';
         /* post it: */
         tinylog_post(tinylog, logline, len);
      }
   }

   /* here on eof */
   /* XXX, ignoring errors now: */
   fchmod(tinylog->fd_current, 0744);
   close(tinylog->fd_current);

   return 0;
}


int main(int argc, char *argv[])
{
   static char    pidbuf[NFMT_SIZE];
   struct tinylog tinylog;
   char           opt;
   nextopt_t      nopt = nextopt_INIT(argc, argv, ":hVk:rs:tz");
   int            opt_resume = 1;
   uint32_t       n = 0;
   const char    *nscan_end;
   int            err = 0;

   /* initialize defaults: */
   tinylog.wantstamp = 0;
   tinylog.wantzip = 0;
   tinylog.current_max = CURRENT_MAX;
   tinylog.keep_max = 5;

   mypid = getpid();
   progpid = nfmt_uint32(pidbuf, (uint32_t) mypid);
   progname = nextopt_progname(&nopt);
   while ((opt = nextopt(&nopt))) {
      char           optc[2] = { nopt.opt_got, '\0' };
      switch (opt) {
      case 'h':
         usage();
         die(0);
         break;
      case 'V':
         version();
         die(0);
         break;
      case 'k':
         n = nscan_uint32(nopt.opt_arg, &nscan_end);
         if (*nscan_end != '\0') {
            fatal_usage("numeric argument required for option -", optc);
         }
         tinylog.keep_max = (size_t) n;
         break;
      case 'r':
         opt_resume = 0;
         break;
      case 's':
         n = nscan_uint32(nopt.opt_arg, &nscan_end);
         if (*nscan_end != '\0') {
            fatal_usage("numeric argument required for option -", optc);
         }
         if (n < (LOGLINE_MAX * 2))
            n = (LOGLINE_MAX * 2);
         tinylog.current_max = (size_t) n;
         break;
      case 't':
         tinylog.wantstamp = 1;
         break;
      case 'z':
         tinylog.wantzip = 1;
         break;
      case ':':
         fatal_usage("missing argument for option -", optc);
         break;
      case '?':
         if (nopt.opt_got != '?') {
            fatal_usage("invalid option -", optc);
         }
         /* else fallthrough: */
      default:
         die_usage();
         break;
      }
   }

   argc -= nopt.arg_ndx;
   argv += nopt.arg_ndx;

   if (argc < 1) {
      fatal_usage("missing log directory argument");
   }

   /* log output directory: */
   tinylog.fn_logdir = argv[0];
   log_info("starting for logging in ", tinylog.fn_logdir, " ...");

   /* where to find gzip: */
   if (tinylog.wantzip) {
      gzip_path = getenv("TINYLOG_ZIP");
      if ((gzip_path == NULL) || (gzip_path[0] == '\0')) {
         /* tinylog.h: */
         gzip_path = TINYLOG_ZIP;
      }
   }

   if (chdir(".") == -1) {
      fatal_syserr("fail chdir() on startup directory");
   }
   if ((tinylog.fd_orig = open(".", O_RDONLY)) == -1) {
      fatal_syserr("fail open() on startup directory");
   }
   fd_cloexec(tinylog.fd_orig);

   /* initialize signal set: */
   sigset_empty(&my_sigset);
   sigset_add(&my_sigset, SIGTERM);
   sigset_add(&my_sigset, SIGHUP);
   /* block signals: */
   sigset_block(&my_sigset);
   /* install signal handlers
    ** note:
    **   sig_catch() does _not_ set SA_RESTART sigaction flag
    **   so calls like read() will be interrupted errno = EINTR on signal
    **   which is what we want, esp SIGTERM during blocking read() on stdin
    */
   sig_catch(SIGTERM, &sig_handler);
   sig_catch(SIGHUP, &sig_handler);

   /* open and cd to logdir: */
   init_logdir(&tinylog);

   /* open/reopen current: */
   init_current(&tinylog, opt_resume);

   /* start logging: */
   sigset_unblock(&my_sigset);
   err = do_log(&tinylog);

   return (err ? 111 : 0);
}

/* eof: tinylog.c */
